js/src/modals/Transfer/store.js

```diff
@@ -351,76 +353,215 @@ export default class TransferStore {
+          eth: gasTotal,
+          token: value
+        };
+      }
+
+      // Otherwise, eth is the sum of the gas and the user input
+      const totalEthValue = gasTotal.plus(value);
+
+      return {
+        eth: totalEthValue,
+        token: value
+      };
+    }
+
+    // If it's the total balance that needs to be sent, send the total balance
+    // if it's not a proper ETH transfer
+    if (!isEth || isWallet) {
+      const tokenBalance = this.getTokenBalance();
+
+      return {
+        eth: gasTotal,
+        token: tokenBalance
+      };
+    }
+
+    // Otherwise, substract the gas estimate
+    const availableEth = this.getTokenBalance('ETH');
+    const totalEthValue = availableEth.gt(gasTotal)
+      ? availableEth.minus(gasTotal)
+      : new BigNumber(0);
+
+    return {
+      eth: totalEthValue.plus(gasTotal),
+      token: totalEthValue
+    };
+  }
+
+  getFormattedTokenValue (tokenValue) {
+    const token = this.getToken();
+
+    if (!token) {
+      return new BigNumber(0);
+    }
+
+    const tag = token.token && token.token.tag || '';
+
+    return this.getTokenValue(tag, tokenValue, true);
+  }
+
+  @action recalculate = (redo = false) => {
+    const { account } = this;
+
+    if (!account || !this.balance) {
+      return;
+    }
+
+    const balance = this.getBalance();
+
+    if (!balance) {
+      return;
     }
 
-    if (totalEth.gt(availableEth)) {
+    const gasTotal = new BigNumber(this.gasStore.price || 0).mul(new BigNumber(this.gasStore.gas || 0));
+
+    const ethBalance = this.getTokenBalance('ETH', true);
+    const tokenBalance = this.getTokenBalance();
+    const { eth, token } = this.getValues(gasTotal);
+
+    let totalEth = gasTotal;
+    let totalError = null;
+    let valueError = null;
+
+    if (eth.gt(ethBalance)) {
       totalError = ERRORS.largeAmount;
     }
 
+    if (token && token.gt(tokenBalance)) {
+      valueError = ERRORS.largeAmount;
+    }
+
+    log.debug('@recalculate', {
+      eth: eth.toFormat(),
+      token: token.toFormat(),
+      ethBalance: ethBalance.toFormat(),
+      tokenBalance: tokenBalance.toFormat(),
+      gasTotal: gasTotal.toFormat()
+    });
+
     transaction(() => {
-      this.total = this.api.util.fromWei(totalEth).toFixed();
       this.totalError = totalError;
-      this.value = value;
       this.valueError = valueError;
       this.gasStore.setErrorTotal(totalError);
       this.gasStore.setEthValue(totalEth);
+
+      this.total = this.api.util.fromWei(eth).toFixed();
+
+      const nextValue = this.getFormattedTokenValue(token);
+      let prevValue;
+
+      try {
+        prevValue = new BigNumber(this.value || 0);
+      } catch (error) {
+        prevValue = new BigNumber(0);
+      }
+
+      // Change the input only if necessary
+      if (!nextValue.eq(prevValue)) {
+        this.value = nextValue.toString();
+      }
+
+      // Re Calculate gas once more to be sure
+      if (redo) {
+        return this.recalculateGas(false);
```

Comment:
The logic is that : the gas estimation is based on the previous value, which can be modified in the method. Let's say you want to send 5 TOK. You estimate gas with the current value (which might be 0), then change to value accordingly (eg. for full balance you need gas estimation before changing the value). But sending 5 TOK costs more gas than sending 0. Thus you might want to recompute the gas estimate.

From what I understand, this is only needed for full balance transfers ; however I might be wrong and that's why I kept it like this. Any thoughts?